🎮Implement the functionality to create and edit blog posts

OlenaYefymenko opened the pull request:
This PR:

- Adds new functions for the creation and edition of the posts;
- Integrates the tests;
- Provides updated templates and icons for the implementation of new functionality.

Link to task https://tutorial.djangogirls.org/en/django_forms/

Ref #18

OlenaYefymenko committed:
Integrate the new functionality for the creation new posts

OlenaYefymenko committed:
Integrate the function to implement the edition of the posts

OlenaYefymenko committed:
Add user authentication check for post creation and editing

OlenaYefymenko committed:
Add tests for the new function providing the creation of new posts

OlenaYefymenko committed:
Implement user authentication, it will be squash to the previous commit

webknjaz commented on blog/views.py:
```diff
@@ -20,3 +22,36 @@ def post_detail(request, pk):
+            post.save()
+            return redirect('post_detail', pk=post.pk)
+    else:
+        form = PostForm
```
Did you forget to initialize the object here?

OlenaYefymenko replied:
Yes

webknjaz replied:
marking as resolved

OlenaYefymenko committed:
Edit typo in `post_new` function

OlenaYefymenko committed:
Delete the redundant check in the test

OlenaYefymenko committed:
Add additional test checks for block `else` in the `post_new` function

OlenaYefymenko committed:
Add draft tests

webknjaz commented on blog/templates/blog/base.html:
```diff
@@ -9,15 +9,20 @@
                 <div class="col">
-                {% block content %}
-                {% endblock %}
+                    {% block content %}
+                    {% endblock %}
```
there's no need to indent this

OlenaYefymenko replied:
I will change indent in the separate PR.

webknjaz replied:
why?

OlenaYefymenko replied:
I fixed the extra indents.

webknjaz replied:
resolved.

webknjaz commented on blog/templates/registration/login.html:
```diff
@@ -0,0 +1,26 @@
+{% block content %}
+
+  {% if form.errors %}
+    <p>Your username and password didn't match. Please try again.
```
```suggestion
    <p>Your username and password didn't match. Please try again.</p>
```

webknjaz commented on blog/urls.py:
```diff
@@ -1,10 +1,15 @@
+    path('post/new/', views.post_new, name='post_new'),
+    path('post/<int:pk>/edit/', views.post_edit, name='post_edit'),
+    path('accounts/login/', auth_views.LoginView.as_view(), name='login'),
+
```
```suggestion
```

webknjaz commented on blog/templates/blog/base.html:
```diff
@@ -9,15 +9,20 @@
-          <div class="container">
-              <h1><a href="/">Django Girls Blog</a></h1>
-          </div>
+            <div class="container">
```
Why did you change the indentation here?

OlenaYefymenko replied:
Edited

webknjaz replied:
What did you edit exactly and how?

OlenaYefymenko replied:
I'm sorry I didn't understand what I should correct

webknjaz commented on blog/forms.py:
```diff
@@ -0,0 +1,15 @@
+
+
+class PostForm(forms.ModelForm):
+    """Form for the creation of new posts in the Blog application."""
```
```suggestion
    """Form for editing posts in the Blog application."""
```

OlenaYefymenko committed:
Update blog/templates/registration/login.html

Co-authored-by: Sviatoslav Sydorenko (Святослав Сидоренко) <wk.cvs.github@sydorenko.org.ua>

OlenaYefymenko committed:
Update blog/urls.py

Co-authored-by: Sviatoslav Sydorenko (Святослав Сидоренко) <wk.cvs.github@sydorenko.org.ua>

OlenaYefymenko committed:
Update blog/forms.py

Co-authored-by: Sviatoslav Sydorenko (Святослав Сидоренко) <wk.cvs.github@sydorenko.org.ua>

OlenaYefymenko committed:
Correct base httm template

OlenaYefymenko committed:
Edit the tests

Edit tests

OlenaYefymenko commented:
@webknjaz in this PR a lot of non-informative commits, I'll do a squash in the last push.

OlenaYefymenko committed:
Add the docstrings

OlenaYefymenko committed:
Edit the tests and templates

This patch provides property tests after the last review and includes
additional logic for user login and logout.

Tests

OlenaYefymenko commented:
```
django-girls-blog-OlenaEfymenko on  18-12-django-forms-new-edit via 🐍 3.11.2 using ☁️  default/fertu-pipeline-dev via django-girls-blog-OlenaEfymenko took 2.2s 
➜ python -m coverage run manage.py test      
Traceback (most recent call last):
  File "/home/olenayefymenko/django-girls-blog-OlenaEfymenko/manage.py", line 73, in <module>
    main()
  File "/home/olenayefymenko/.pyenv/versions/3.11.2/lib/python3.11/contextlib.py", line 81, in inner
    return func(*args, **kwds)
           ^^^^^^^^^^^^^^^^^^^
  File "/home/olenayefymenko/django-girls-blog-OlenaEfymenko/manage.py", line 69, in main
    execute_from_command_line(sys.argv)
  File "/home/olenayefymenko/django-girls-blog-OlenaEfymenko/.venv/lib/python3.11/site-packages/django/core/management/__init__.py", line 442, in execute_from_command_line
    utility.execute()
  File "/home/olenayefymenko/django-girls-blog-OlenaEfymenko/.venv/lib/python3.11/site-packages/django/core/management/__init__.py", line 436, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/olenayefymenko/django-girls-blog-OlenaEfymenko/.venv/lib/python3.11/site-packages/django/core/management/__init__.py", line 275, in fetch_command
    klass = load_command_class(app_name, subcommand)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/olenayefymenko/django-girls-blog-OlenaEfymenko/.venv/lib/python3.11/site-packages/django/core/management/__init__.py", line 48, in load_command_class
    module = import_module("%s.management.commands.%s" % (app_name, name))
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/olenayefymenko/.pyenv/versions/3.11.2/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1206, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1178, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1149, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/home/olenayefymenko/django-girls-blog-OlenaEfymenko/.venv/lib/python3.11/site-packages/django/core/management/commands/test.py", line 6, in <module>
    from django.test.runner import get_max_test_processes
  File "/home/olenayefymenko/django-girls-blog-OlenaEfymenko/.venv/lib/python3.11/site-packages/django/test/__init__.py", line 3, in <module>
    from django.test.client import AsyncClient, AsyncRequestFactory, Client, RequestFactory
  File "/home/olenayefymenko/django-girls-blog-OlenaEfymenko/.venv/lib/python3.11/site-packages/django/test/client.py", line 727
    session = await self.asession()
              ^^^^^^^^^^^^^^^^^^^^^
SyntaxError: 'await' outside async function
/home/olenayefymenko/django-girls-blog-OlenaEfymenko/.venv/lib/python3.11/site-packages/coverage/inorout.py:519: CoverageWarning: Module blog was previously imported, but not measured (module-not-measured)
  self.warn(msg, slug="module-not-measured")
/home/olenayefymenko/django-girls-blog-OlenaEfymenko/.venv/lib/python3.11/site-packages/coverage/inorout.py:519: CoverageWarning: Module mysite was previously imported, but not measured (module-not-measured)
  self.warn(msg, slug="module-not-measured")
/home/olenayefymenko/django-girls-blog-OlenaEfymenko/.venv/lib/python3.11/site-packages/coverage/control.py:887: CoverageWarning: No data was collected. (no-data-collected)
  self._warn("No data was collected.", slug="no-data-collected")
```

webknjaz commented on blog/forms.py:
```diff
@@ -0,0 +1,15 @@
+"""Form class for the Blog application."""
```
This is not a class but a module.

OlenaYefymenko replied:
Edited

webknjaz replied:
commented

webknjaz commented on mysite/urls.py:
```diff
@@ -1,9 +1,16 @@
+        name='logout',
+    ),
     path('', include('blog.urls')),
+
```
what's the purpose of an extra empty line here?

OlenaYefymenko replied:
Removed

webknjaz commented on tests/test_views.py:
```diff
@@ -16,7 +16,10 @@ class ViewsTest(TestCase):
     def setUpTestData(cls):
         """Create test client, user, posts with different published dates."""
-        cls.user = get_user_model().objects.create()
+        cls.user = get_user_model().objects.create_user(
+            username='test user',
+            password='test password',
+        )
```
Is this change actually necessary for the tests to work?

OlenaYefymenko replied:
I deleted it for now

webknjaz replied:
resolved.

webknjaz commented on tests/test_views.py:
```diff
@@ -86,3 +99,103 @@ def test_post_detail_failed(self):
+        self.assertEqual(Post.objects.count(), initial_post_count + 1)
+
+    def test_post_new_invalid_form_authorized(self):
+        """Add."""
```
huh?

webknjaz replied:
marking as resolved

OlenaYefymenko commented:
@webknjaz thank you for review, I'll change and push

OlenaYefymenko committed:
Update dependencies to fix test runner async error

OlenaYefymenko committed:
Delete the extra line

webknjaz commented on blog/templates/blog/base.html:
```diff
@@ -9,11 +9,19 @@
+            </div>
         </header>
-        <main class="container">
+        <main class="content container">
```
How does this relate to editing the posts?

webknjaz commented on requirements/dev.txt:
Perhaps perform package bumps in a separate PR?

OlenaYefymenko replied:
yep, this PR:
https://github.com/kpi-web-guild/django-girls-blog-OlenaEfymenko/pull/45

webknjaz commented on tests/test_views.py:
```diff
@@ -86,3 +96,103 @@ def test_post_detail_failed(self):
+                kwargs={'pk': new_post.pk},
+            ),
+        )
+        self.assertEqual(new_post.title, 'Test Title')
```
It's a bit difficult to connect the dots when this constant string is defined here, but the data it compares to is defined elsewhere. Why not reuse the same value saved into a var?

OlenaYefymenko replied:
0970a1bc13850a62f0599722a7dea95139c3e529

webknjaz commented on tests/test_views.py:
```diff
@@ -86,3 +96,103 @@ def test_post_detail_failed(self):
+        http_response = self.client.post(url, self.invalid_form_post)
+        self.assertEqual(http_response.status_code, 200)
+        self.assertTemplateUsed(http_response, 'blog/post_edit.html')
+        self.assertContains(
+            http_response, 'This field is required.', html=True,
+        )
```
Try using the `assertFormError()` helper:
- https://docs.djangoproject.com/en/5.1/topics/testing/tools/#django.test.SimpleTestCase.assertFormError
- https://developer.mozilla.org/en-US/docs/Learn/Server-side/Django/Testing#testing_views_with_forms
- https://gist.github.com/cwurld/f132f673446158808952#file-django_testcase_cheatsheet-py-L80

OlenaYefymenko replied:
3b2597b14c08d3a8eaea1e8453b692f8bf252361

webknjaz commented on tests/test_views.py:
```diff
@@ -86,3 +96,103 @@ def test_post_detail_failed(self):
+        self.assertEqual(post.title, 'Updated Title')
+        self.assertEqual(post.text, 'Updated Text')
+        self.assertEqual(post.author, self.user)
+        self.assertNotEqual(post.published_date, initial_published_date)
```
While this is a useful check, the best way of dealing with the time changes is freezing it. There are libraries like freezegun and similar that monkey-patch the stdlib that could be used. But since you did it manually in the past already, I suggest repeating the same technique here: https://github.com/kpi-web-guild/django-girls-blog-OlenaEfymenko/blob/f6ccd09/tests/test_views.py#L60-L63.

OlenaYefymenko replied:
75470d25c3479b6906c5f0ac73dfe49bd7ee8a64

OlenaYefymenko commented on tests/test_views.py:
```diff
@@ -86,3 +96,103 @@ def test_post_detail_failed(self):
+        self.assertNotEqual(post.published_date, initial_published_date)
+
+    def test_post_edit_invalid_form_authorized(self):
+        """Test handling of invalid form submission when editing a post."""
```
99cfc585264a9587b9b15e64f7ad42f0171dbcee

webknjaz reviewed (approved):
blanket
